featuregates/controller/pkg/feature/feature_controller.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package feature
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/vmware-tanzu/tanzu-framework/featuregates/client/pkg/util"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	corev1alpha2 "github.com/vmware-tanzu/tanzu-framework/apis/core/v1alpha2"
+)
+
+const contextTimeout = 30 * time.Second
+
+// FeatureReconciler reconciles a Feature object.
+type FeatureReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core.tanzu.vmware.com,resources=featuregates,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=core.tanzu.vmware.com,resources=featuregates/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core.tanzu.vmware.com,resources=features,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core.tanzu.vmware.com,resources=features/status,verbs=get;update;patch
+
+// Reconcile reconciles the FeatureGate spec by computing activated, deactivated and unavailable features.
+func (r *FeatureReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctxCancel, cancel := context.WithTimeout(ctx, contextTimeout)
+	defer cancel()
+
+	log := r.Log.WithValues("feature", req.NamespacedName)
+	log.Info("Starting reconcile")
+
+	feature := &corev1alpha2.Feature{}
+	if err := r.Client.Get(ctxCancel, req.NamespacedName, feature); err != nil {
+		// Reconcile deleted feature
```

Comment:
Could you expand these comments? I don't think things are obvious in this controller, so the more detailed comments the better.

Reply:
+1

This controller is doing some things that are not obvious to wrap my head around. So comments that quickly tell us what the goal of "reconcile deleted feature" is would be very helpful.

Reply:
not sure if this [comment](https://github.com/vmware-tanzu/tanzu-framework/pull/3613/files#diff-2f55534c92afa48c0492aa91850834888c7c5fce2af6cf6a661137cf84d1f631R102) is helpful, but, yes i can expand more

Reply:
> Check if feature part of any FeatureGate spec and update status in Feature result to Invalid

That comment is confusing. Do you mean,

Check if the features listed in the FeatureGate spec FeatureReference slice is invalid and update the Feature result appropriately.

Might explain for line 52, "Reconcile deleted feature" what it means to be invalid due to deletion?

Reply:
The controller is reconciling on the Feature resource, so it is basically saying when reconciling the deleted feature, check if it is part of any FeatureGate spec and update the Feature result to invalid in FeatureGate status